File: src/main/scala/NextClosestTime.scala
object NextClosestTime extends App {

  def nextClosestTime(time: String): String = {
    if (time == null || time.isEmpty) time
    else {
      var cur = time.substring(0,2).toInt * 60 + time.substring(3).toInt
      val allowed = time.filter(_ != ':').map( _.toInt - '0').toSet
      while(true){
        cur  = (cur + 1) % (24 * 60)
        val digits = List(cur/60 / 10, cur / 60 % 10, (cur % 60) / 10,(cur % 60) % 10 )
        var i = 0
        var error = false
        while(i < digits.size && !error){
          if(!allowed.contains(digits(i)))
            error = true
          i+= 1
        }
        if(!error) {
          val hours = cur / 60
          val minutes = cur % 60
          return f"$hours%02d:$minutes%02d"
        }
      }
      ""
    }
  }

  println(nextClosestTime("23:59"))
  println(nextClosestTime("19:34"))
}

File: src/main/scala/mock/25072020/NextClosestTime.py
class Solution:
    def nextClosestTime(self, time):
        if not time:
            return time

        cur = 60 * int(time[0:2]) + int(time[3:])
        allowed = {int(c) for c in time if c != ':'}

        while True:
            cur = (cur + 1) % (24 * 60)
            digits = [cur // 60 // 10, cur // 60 % 10, cur % 60 // 10, cur % 60 % 10]
            for digit in digits:
                if all(digit in allowed for digit in digits):
                    return '{}{}:{}{}'.format(digits[0], digits[1], digits[2], digits[3])


sol = Solution()
print(sol.nextClosestTime("23:59"))
print(sol.nextClosestTime("19:34"))
